File: web/gauges/gaugeMath.ts
export interface GaugeRange {
  min: number;
  max: number;
  warnLow?: number;
  warnHigh?: number;
  dangerLow?: number;
  dangerHigh?: number;
}

export type GaugeZone = "ok" | "warn" | "danger";

/** Where `value` sits between `min` and `max`, clamped to [0, 1]. */
export function clampFraction(value: number, min: number, max: number): number {
  if (max <= min) return 0;
  return Math.min(1, Math.max(0, (value - min) / (max - min)));
}

export function zoneFor(value: number, range: GaugeRange): GaugeZone {
  const { warnLow, warnHigh, dangerLow, dangerHigh } = range;
  if ((dangerLow !== undefined && value <= dangerLow) || (dangerHigh !== undefined && value >= dangerHigh)) {
    return "danger";
  }
  if ((warnLow !== undefined && value <= warnLow) || (warnHigh !== undefined && value >= warnHigh)) {
    return "warn";
  }
  return "ok";
}

const ZONE_COLORS: Record<GaugeZone, string> = {
  ok: "#4caf50",
  warn: "#e0c34c",
  danger: "#e55353",
};

export function colorForZone(zone: GaugeZone): string {
  return ZONE_COLORS[zone];
}

File: desktop/ProtonEcuToolkit.App/Gauges/GaugeMath.cs
using System.Windows.Media;

namespace ProtonEcuToolkit.App.Gauges;

public enum GaugeZone
{
    Ok,
    Danger,
}

public static class GaugeMath
{
    /// <summary>Where <paramref name="value"/> sits between min and max, clamped to [0, 1].</summary>
    public static double ClampFraction(double value, double min, double max)
    {
        if (max <= min) return 0;
        return Math.Min(1, Math.Max(0, (value - min) / (max - min)));
    }

    /// <summary>Danger once value reaches the redline (e.g. RPM redline, an overheat threshold, etc). No redline set = always ok.</summary>
    public static GaugeZone ZoneFor(double value, double? redline) =>
        redline is { } r && value >= r ? GaugeZone.Danger : GaugeZone.Ok;

    public static Color ColorForZone(GaugeZone zone) => zone switch
    {
        GaugeZone.Ok => Color.FromRgb(0xf2, 0xf2, 0xf2),
        GaugeZone.Danger => Color.FromRgb(0xe5, 0x35, 0x35),
        _ => throw new ArgumentOutOfRangeException(nameof(zone)),
    };
}
